pkg/controller/disruption/disruption.go

```diff
@@ -176,89 +176,84 @@ func (dc *DisruptionController) finders() []podControllerFinder {
 // getPodReplicaSets finds replicasets which have no matching deployments.
 func (dc *DisruptionController) getPodReplicaSets(pod *v1.Pod) ([]controllerAndScale, error) {
 	cas := []controllerAndScale{}
-	rss, err := dc.rsLister.GetPodReplicaSets(pod)
-	// GetPodReplicaSets returns an error only if no ReplicaSets are found.  We
-	// don't return that as an error to the caller.
-	if err != nil {
-		return cas, nil
-	}
-	controllerScale := map[types.UID]int32{}
-	for _, rs := range rss {
-		// GetDeploymentsForReplicaSet returns an error only if no matching
+	if controllerRef := controller.GetControllerOf(pod); controllerRef != nil {
+		rs, err := dc.rsLister.ReplicaSets(pod.Namespace).Get(controllerRef.Name)
+		if err != nil {
+			return cas, nil
+		}
+		controllerScale := map[types.UID]int32{}
 		// deployments are found.
-		_, err := dc.dLister.GetDeploymentsForReplicaSet(rs)
-		if err == nil { // A deployment was found, so this finder will not count this RS.
-			continue
+		if controllerRef := controller.GetControllerOf(rs); controllerRef != nil {
```

Comment:
I think this may be why the test is failing, at least for RS. If `controllerRef == nil`, we actually want to return the RS without any further checking.

The idea here is, if the RS is a top-level controller, we return it. If the RS is controlled by a Deployment, we *don't* return it, because we instead want the Deployment itself to be returned by `getPodDeployments`. If the RS is controlled by something that isn't a Deployment, we should still return it because we don't know the semantics of that unknown thing.

There's an opportunity now to simplify and optimize this whole process. We can actually get rid of the finders and have one function called something like `getPodControllerScale`. It would check the Pod's ControllerRef and figure out the scale somehow depending on the type of controller. Let me know if you want to pursue that longer-term fix and we can talk about it. It's also fine if you just want to finish this intermediate fix first and leave the refactoring for later.